File: src/forward_model/run_spectrum.zig
const std = @import("std");
const common = @import("radiative_transfer/root.zig");
const measurement = @import("instrument_grid/root.zig");
const Method = @import("method.zig").Method;
const providers = @import("builtins/root.zig");
const Scene = @import("../input/Scene.zig").Scene;
const SummaryWorkspace = @import("instrument_grid/grid_calculation/workspace.zig").SummaryWorkspace;
const PreparedOpticalState = @import("optical_properties/root.zig").PreparedOpticalState;

pub const Result = measurement.MeasurementSpaceProduct;

pub fn run(
    allocator: std.mem.Allocator,
    case: *const Scene,
    optics: *const PreparedOpticalState,
    work: ?*SummaryWorkspace,
    method: Method,
    rtm_controls: common.RadiativeTransferControls,
) !Result {
    switch (method) {
        .exact => {},
    }

    const route = try common.prepareRoute(.{
        .regime = case.observation_model.regime,
        .execution_mode = .scalar,
        .derivative_mode = .none,
        .rtm_controls = rtm_controls,
    });

    if (work) |workspace| {
        const view = try measurement.simulateProductWithWorkspace(
            allocator,
            workspace,
            case,
            route,
            optics,
            providers.exact(),
        );
        return view.toOwned(allocator);
    }

    return measurement.simulateProduct(
        allocator,
        case,
        route,
        optics,
        providers.exact(),
    );
}

File: rust_src/forward_model/run_spectrum.rs
use crate::{
    forward_model::{
        instrument_grid::grid_calculation::{
            simulate::{self, simulate_product},
            types::InstrumentGridProduct,
        },
        method::Method,
        optical_properties::state_build::PreparedOpticalState,
        radiative_transfer::{
            common_route,
            common_types::{self, DispatchRequest, ExecutionMode, RadiativeTransferControls},
        },
    },
    input::scene::Scene,
};

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Error {
    Route(common_types::Error),
    Simulation(simulate::Error),
}

impl From<common_types::Error> for Error {
    fn from(value: common_types::Error) -> Self {
        Self::Route(value)
    }
}

impl From<simulate::Error> for Error {
    fn from(value: simulate::Error) -> Self {
        Self::Simulation(value)
    }
}

pub fn run(
    scene: &Scene,
    optics: &PreparedOpticalState,
    method: Method,
    rtm_controls: RadiativeTransferControls,
) -> Result<InstrumentGridProduct, Error> {
    match method {
        Method::Exact => {}
    }

    let route = common_route::prepare_route(DispatchRequest {
        regime: scene.observation_model.regime,
        execution_mode: ExecutionMode::Scalar,
        derivative_mode: crate::input::scene::DerivativeMode::None,
        rtm_controls,
    })?;
    simulate_product(scene, route, optics).map_err(Error::from)
}
